packages/engine/src/scene/systems/EnvironmentSystem.tsx

```diff
@@ -28,32 +28,54 @@ import React, { useEffect } from 'react'
 import {
   defineSystem,
   Entity,
+  getComponent,
   PresentationSystemGroup,
   QueryReactor,
   useComponent,
-  useEntityContext
+  useEntityContext,
+  useOptionalComponent
 } from '@ir-engine/ecs'
-import { GroupComponent } from '@ir-engine/spatial/src/renderer/components/GroupComponent'
 import { BackgroundComponent } from '@ir-engine/spatial/src/renderer/components/SceneComponents'
-import { haveCommonAncestor } from '@ir-engine/spatial/src/transform/components/EntityTree'
+import { haveCommonAncestor, useChildrenWithComponents } from '@ir-engine/spatial/src/transform/components/EntityTree'
 
+import { MeshComponent } from '@ir-engine/spatial/src/renderer/components/MeshComponent'
+import { VisibleComponent } from '@ir-engine/spatial/src/renderer/components/VisibleComponent'
 import { EnvmapComponent, updateEnvMap } from '../components/EnvmapComponent'
+import { SourceComponent } from '../components/SourceComponent'
 import { EnvMapSourceType } from '../constants/EnvMapEnum'
 
 const EnvmapReactor = (props: { backgroundEntity: Entity }) => {
   const entity = useEntityContext()
   const envmapComponent = useComponent(entity, EnvmapComponent)
   const backgroundComponent = useComponent(props.backgroundEntity, BackgroundComponent)
-  const groupComponent = useComponent(entity, GroupComponent)
+  const hasRootMesh = !!useOptionalComponent(entity, MeshComponent)
+  const childrenMesh = useChildrenWithComponents(
+    entity,
+    [MeshComponent, VisibleComponent, SourceComponent],
+    [EnvmapComponent]
+  )
+
+  const getMeshes = () => {
+    const meshEntities = [...childrenMesh]
+    if (hasRootMesh) meshEntities.push(entity)
+
+    return meshEntities.map((meshEntity) => getComponent(meshEntity, MeshComponent))
+  }
 
   useEffect(() => {
-    // TODO use spatial queries
     if (!haveCommonAncestor(entity, props.backgroundEntity)) return
     if (envmapComponent.type.value !== EnvMapSourceType.Skybox) return
-    for (const obj of groupComponent.value) {
-      updateEnvMap(obj as any, backgroundComponent.value as any)
+    const meshes = getMeshes()
+
+    for (const mesh of meshes) {
+      updateEnvMap(mesh, backgroundComponent.value as any)
+    }
+    return () => {
+      for (const mesh of meshes) {
+        updateEnvMap(mesh, null)
+      }
     }
-  }, [envmapComponent.type, backgroundComponent])
+  }, [childrenMesh.join(), envmapComponent.type, backgroundComponent])
```

Comment:
the .join() here is redundant, what is returned from useChildrenWithComponents will be correctly reactive